test/test_html5_input.c

```diff
@@ -0,0 +1,348 @@
+    printf("    FAIL    unexpected call to %s\n", __func__);
+    return false;
+  }
+
+  if (strcmp(inputEvent->inputType, inputType_insertFromDrop) != 0) {
+    printf("    FAIL    inputType is not '%s': %s\n",
+           inputType_insertFromDrop,
+           inputEvent->inputType);
+    return false;
+  }
+
+  if (strcmp(inputEvent->data, longDropText) != 0) {
+    printf("    FAIL    wrong text received: %s\n", inputEvent->data);
+    return false;
+  }
+
+  printf("    SUCCESS\n");
+  goalPassed_dropVeryLongText = true;
+  startTestChangeCheckbox();
+  return true;
+}
+
+static void startTestDropLongText() {
+  currentGoal = GoalDragAndDropVeryLongText;
+  emscripten_html5_remove_all_event_listeners();
+  const EMSCRIPTEN_RESULT res = emscripten_set_input_callback(
+    "#input-a", 0, false, inputCallbackWaitingForPasteOfLongText);
+  ASSERT_RESULT(res,
+                EMSCRIPTEN_RESULT_SUCCESS,
+                "install callback on existing input element");
+
+  setInputAValue("");
+  setInputBValue(longDropText);
+  printf("Please drag&drop all text from B to A (Ctrl+A, drag, drop).\n");
+}
+
+static const char* const expectedPasteText = "I like emscripten.";
+
+static bool
+inputCallbackWaitingForPaste(const int eventType,
+                             const EmscriptenInputEvent* const inputEvent
+                             __attribute__((nonnull)),
+                             void* const userData) {
+  printEvent(inputEvent, __func__);
+
+  if (currentGoal != GoalPasteText) {
+    printf("    FAIL    unexpected call to %s\n", __func__);
+    return false;
+  }
+
+  if (strcmp(inputEvent->inputType, inputType_insertFromPaste) != 0) {
+    printf("    FAIL    inputType is not '%s': %s\n",
+           inputType_insertFromPaste,
+           inputEvent->inputType);
+    return false;
+  }
+
+  if (strcmp(inputEvent->data, expectedPasteText) != 0) {
+    printf("    FAIL    wrong text received: %s\n", inputEvent->data);
+    return false;
+  }
+
+  printf("    SUCCESS\n");
+  goalPassed_pasteText = true;
+  startTestDropLongText();
+  return true;
+}
+
+static void startTestPasteText() {
+  currentGoal = GoalPasteText;
+  emscripten_html5_remove_all_event_listeners();
+  const EMSCRIPTEN_RESULT res = emscripten_set_input_callback(
+    "#input-b", 0, false, inputCallbackWaitingForPaste);
+  ASSERT_RESULT(res,
+                EMSCRIPTEN_RESULT_SUCCESS,
+                "install callback on existing input element");
+
+  setInputAValue(expectedPasteText);
+  printf("Please copy&paste all text from A to B (Ctrl+A, Ctrl+C, Ctrl+V).\n");
+}
+
+static void* const expectedUserDataPtr = (void*)0x13370001;
+
+static bool
+inputCallbackWaitingForLetterA(const int eventType,
+                               const EmscriptenInputEvent* const inputEvent
+                               __attribute__((nonnull)),
+                               void* const userData) {
+  printEvent(inputEvent, __func__);
+
+  if (currentGoal != GoalInputSingleLetter) {
```

Comment:
How about just `assert(currentGoal == GoalInputSingleLetter)` here and elsewhere?